### ISSUE TITLE
logstash-input-azureblob : There is already a lease present

### ISSUE THREAD

**Issue by EmFl** [Azure/azure-diagnostics-tools#92]
Hello,
We're getting rather frequently the following error : 

```
{
:timestamp=>"2017-09-12T08:32:49.026000+0000",
:message=>"Oh My, An error occurred. LeaseAlreadyPresent (409): There is already a lease present.\nRequestId:xxxxxxxxxxxx\nTime:2017-09-12T08:32:49.3075152Z:\n[\"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/http/retry_policy.rb:48:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/http/http_request.rb:104:in `with_filter'\", \"org/jruby/RubyMethod.java:116:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/http/signer_filter.rb:28:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/http/http_request.rb:104:in `with_filter'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/service.rb:36:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/filtered_service.rb:34:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-core-0.1.11/lib/azure/core/signed_service.rb:41:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-storage-0.12.3.preview/lib/azure/storage/service/storage_service.rb:53:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-storage-0.12.3.preview/lib/azure/storage/blob/blob_service.rb:59:in `call'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-storage-0.12.3.preview/lib/azure/storage/blob/blob_service.rb:190:in `acquire_lease'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/azure-storage-0.12.3.preview/lib/azure/storage/blob/blob.rb:413:in `acquire_blob_lease'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-azureblob-0.9.9/lib/logstash/inputs/azureblob.rb:309:in `acquire_lease'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-azureblob-0.9.9/lib/logstash/inputs/azureblob.rb:339:in `register_for_read'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-azureblob-0.9.9/lib/logstash/inputs/azureblob.rb:325:in `register_for_read'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-azureblob-0.9.9/lib/logstash/inputs/azureblob.rb:135:in `process'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-input-azureblob-0.9.9/lib/logstash/inputs/azureblob.rb:122:in `run'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.1-java/lib/logstash/pipeline.rb:342:in `inputworker'\", \"/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.4.1-java/lib/logstash/pipeline.rb:336:in `start_input'\"]", :exception=>#<Azure::Core::Http::HTTPError:16009592 @status_code: 409, @http_response: #<Azure::Core::Http::HttpResponse:0x5936850d @http_response=#<Faraday::Response:0x2745c853 @on_complete_callbacks=[], @env=#<Faraday::Env @method=:put @body="\xEF\xBB\xBF<?xml version=\"1.0\" encoding=\"utf-8\"?><Error><Code>LeaseAlreadyPresent</Code><Message>There is already a lease present.\nRequestId:xxxxxxxxxxxx\nTime:2017-09-12T08:32:49.3075152Z</Message></Error>" @url=#<URI::HTTPS:0x7d0e5d4e URL:https://xxxxxxxxxxxx.blob.core.windows.net/xxxxxxxxxxxx-logs/data/registry.lock?comp=lease&timeout=10> @request=#<Faraday::RequestOptions open_timeout=60> @request_headers={"User-Agent"=>"Azure-Storage/0.12.3-preview (Ruby 1.9.3-p551; Linux linux)", "x-ms-date"=>"Tue, 12 Sep 2017 08:32:48 GMT", "x-ms-version"=>"2015-04-05", "DataServiceVersion"=>"1.0;NetFx", "MaxDataServiceVersion"=>"2.0;NetFx", "Content-Type"=>"application/atom+xml; charset=utf-8", "x-ms-lease-action"=>"acquire", "x-ms-lease-duration"=>"-1", "Content-Length"=>"0", "Authorization"=>"SharedKey xxxxxxxxxxxx:xxxxxxxxxxxx"} @ssl=#<Faraday::SSLOptions verify=true> @response=#<Faraday::Response:0x2745c853 ...> @response_headers={"content-length"=>"221", "content-type"=>"application/xml", "server"=>"Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0", "x-ms-request-id"=>"xxxxxxxxxxxx", "x-ms-version"=>"2015-04-05", "date"=>"Tue, 12 Sep 2017 08:32:48 GMT", "connection"=>"close"} @status=409>>, @uri=#<URI::HTTPS:0xa31fff9 URL:https://xxxxxxxxxxxx.blob.core.windows.net/xxxxxxxxxxxx-logs/data/registry.lock?comp=lease&timeout=10>>, @uri: #<URI::HTTPS:0xa31fff9 URL:https://xxxxxxxxxxxx.blob.core.windows.net/xxxxxxxxxxxx-logs/data/registry.lock?comp=lease&timeout=10>, @description: "There is already a lease present.\nRequestId:xxxxxxxxxxxx\nTime:2017-09-12T08:32:49.3075152Z", @type: "LeaseAlreadyPresent">, :level=>:error}
```

Looking at the code I'm not sure I understand exactly when we lock and how it is supposed to work with multiple logstash nodes (we only have one at the moment). It seems that the lock is taken when logstash creates the list of blobs to parse but then it is released. What would prevent a second logstash from then building another list of blobs that would need to be read and reread the same blobs from the other logstash nodes ? 

In any case, this issue is because sometimes the lock isn't removed, I don't know yet when but it happens somehow. This results in logstash not processing any logs from the storage container.

Let me know how can I help debug this, this is a show stopper for us at the moment
Thanks

**Comment by xiaomi7732** [comment]
@EmFl,
Thank you for reporting this issue. I will investigate what is going on. At the same time, please break the lease on the registry file to unblock your scenario.

Thanks,
Saar

**Comment by xiaomi7732** [comment]
@EmFI,
I put in a quick fix in 0.9.10 here: https://rubygems.org/gems/logstash-input-azureblob.
Basically, the issue might still happen due to logstash crash or so, but the logstash client will recover from it and keep getting data afterwards.
#87 will be used to track the long term fix.
